File: API_ThuVien/API_ThuVien/Models/Sach.cs
using System;
using System.Collections.Generic;

namespace API_ThuVien.Models;

public partial class Sach
{
    public int Masach { get; set; }

    public string Tensach { get; set; } = null!;

    public int Matg { get; set; }

    public int Manxb { get; set; }

    public string? Hinhanh { get; set; }

    public string? Theloai { get; set; }

    public string? Mota { get; set; }

    public decimal Giamuon { get; set; }

    public int Soluongton { get; set; }

    public string Trangthai { get; set; } = null!;

    public virtual ICollection<Chitietphieumuon> Chitietphieumuons { get; set; } = new List<Chitietphieumuon>();

    public virtual ICollection<Chitietphieunhap> Chitietphieunhaps { get; set; } = new List<Chitietphieunhap>();

    public virtual ICollection<Chitietphieutra> Chitietphieutras { get; set; } = new List<Chitietphieutra>();

    public virtual ICollection<Chitietthanhly> Chitietthanhlies { get; set; } = new List<Chitietthanhly>();

    [System.Text.Json.Serialization.JsonIgnore] // Thêm dòng này để tránh lỗi vòng lặp JSON khi trả về
    public virtual Nhaxuatban? ManxbNavigation { get; set; }

    [System.Text.Json.Serialization.JsonIgnore]
    public virtual Tacgium? MatgNavigation { get; set; }
}

File: thuvienapp/lib/models/sach.dart
class Sach {
  final int masach;
  final String tensach;
  final String? hinhanh;
  final double giamuon;
  final int soluongton;
  final String? theLoai;
  final String? tenTacGia;
  final String? moTa;
  final String? tennxb;
  final int matg;
  final int manxb;
  final String trangThai;

  Sach({
    required this.masach,
    required this.tensach,
    this.hinhanh,
    required this.giamuon,
    required this.soluongton,
    this.theLoai,
    this.tenTacGia,
    this.moTa,
    this.tennxb,
    required this.matg,
    required this.manxb,
    required this.trangThai,
  });

  factory Sach.fromJson(Map<String, dynamic> json) {
    return Sach(
      masach: json['masach'] ?? 0,
      tensach: json['tensach'] ?? "Sách chưa cập nhật tên",
      hinhanh: json['hinhanh'],
      giamuon: (json['giamuon'] ?? 0).toDouble(),
      soluongton: json['soluongton'] ?? 0,
      // Map đúng key từ Backend trả về ở Bước 1
      theLoai: json['theloai'] ?? "Khác",
      tenTacGia: json['tenTacGia'] ?? "Chưa rõ",
      moTa: json['mota'] ?? "",
      matg: json['matg'] ?? json['Matg'] ?? 0,
      manxb: json['manxb'] ?? json['Manxb'] ?? 0,
      trangThai: json['trangthai'] ?? json['Trangthai'] ?? "Có sẵn",
      tennxb: json['tenNxb'] ?? json['TenNxb'] ?? "Đang cập nhật",
    );
  }
}